File: Explore/Learn/Introduction to Data Structure/Binary Tree/Binary Tree Preorder Traversal/PreorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> res = new ArrayList<>();
    
    public void aux(TreeNode node) {
        if (node == null) return;
        res.add(node.val);
        aux(node.left);
        aux(node.right);
    }
    
    public List<Integer> preorderTraversal(TreeNode root) {
        aux(root);
        return res;
    }
}

File: Explore/Learn/Introduction to Data Structure/Binary Tree/Binary Tree Preorder Traversal/PreorderTraversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    val res = mutableListOf<Int>()
    
    fun aux(node: TreeNode?) {
        if (node == null) return;
        res.add(node.`val`)
        aux(node.left)
        aux(node.right)
    }
    
    fun preorderTraversal(root: TreeNode?): List<Int> {
        aux(root)
        return res
    }
}

